[v4] Dev docs updates to policies (#451)
strapi/documentation: docs/developer-docs/latest/development/backend-customization/policies.md

---
title: Policies - Backend customization - Strapi Developer Documentation
description: …
sidebarDepth: 3
---

<!-- TODO: update SEO -->

# Policies

Policies are functions that execute specific logic on each request before it reaches the [controller](/developer-docs/latest/development/backend-customization/controllers.md). They are mostly used for securing business logic.
<!-- TODO: remove this comment — the link above will work only when merged with PR #446 -->

> Each [route](/developer-docs/latest/development/backend-customization/routes.md) of a Strapi project can be associated to an array of policies. For example, a policy named `is-admin` could check that the request is sent by an admin user, and uses it for critical routes.

Reviewer: ```suggestion
Each [route](/developer-docs/latest/development/backend-customization/routes.md) of a Strapi project can be associated to an array of policies. For example, a policy named `is-admin` could check that the request is sent by an admin user, and restrict access to critical routes.
```

A new policy can be implemented:

- with the [interactive CLI command `strapi generate`](/developer-docs/latest/developer-resources/cli/CLI.md#strapi-generate) 
- or manually by creating a JavaScript file in the appropriate folder (see [project structure](/developer-docs/latest/setup-deployment-guides/file-structure.md)):
  - `./src/policies/` for global policies
  - `./src/api/[api-name]/policies/` for API policies
  - `./src/plugins/[plugin-name]/policies/` for plugin policies

<br/>
Global policy implementation example:

```js
// path: ./src/policies/is-authenticated.js

module.exports = (policyContext, { strapi }) => {
  if (ctx.state.user) { // if a session is open
    // go to next policy or reach the controller's action
    return await next();
  }

  ctx.unauthorized(`You're not logged in!`); // returns a 401 error

  return true; // If you return nothing, Strapi considers you didn't want to block the request and will let it pass
};
```

`policyContext` is a wrapper arround the [controller](/developer-docs/latest/development/backend-customization/controllers.md) context. It adds some logic that can be useful to implement a policy for both REST and GraphQL.

<br/>

> Policies can declare a `config` object to inject some custom configuration.

Reviewer: Policies do not declare config but can be configure with a config

Author: ```suggestion
Policies can be configured using a `config` object.
```

    }
  ]
}
```

### Plugin policies

[Plugins](/developer-docs/latest/plugins/plugins-intro.md) can add and expose policies to an application. For example, the [Users & Permissions plugin](/user-docs/latest/plugins/strapi-plugins.html#users-permissions-plugin) comes with policies to ensure that the user is authenticated or has the rights to perform an action:

```js
// path: ./src/api/restaurant/routes/router.js

module.exports = {
  routes: [
    {
      method: 'GET',
      path: '/restaurants',
      handler: 'Restaurant.find',
      config: {
        /**
          The `isAuthenticated` policy prodived with the `users-permissions` plugin 
          is executed before the `find` action in the `Restaurant.js` controller.
        */
        policies: ['plugins::users-permissions.isAuthenticated']
      }
    }
  ]
}
```

### API policies

API policies are associated to the routes defined in the API where they have been declared.

```js

// path: ./src/api/restaurant/policies/is-admin.js.

module.exports = async (ctx, next) => {
  if (ctx.state.user.role.name === 'Administrator') {
    // Go to next policy or will reach the controller's action.
    return await next();
  }

>   ctx.unauthorized(`You're not allowed to perform this action!`);

Reviewer: same here it should return false or throw an error
